Modify `:contains`, `:starts-with` and `:ends-with` so field or expression can be used as the second argument

lbrdnk opened the pull request:
Closes #31284

### Description

This PR updates compilation process of string filtering expressions `:contains`, `:starts-with` and `:ends-with`. _Field or expression can be used as their second argument_ now for `:sql` and `:mongo` drivers. On the other hand, `:druid`'s code was updated to throw specific exception while attempting _that_ as the filters that are in place support only literal patterns.

~~Another notable change is, that when case is ignored for a match, not only `field`'s (first arg) case is lowered on db level, but also the second `arg`'s. Second arg is wrapped to `:lower`, even if it is literal pattern. Reason for this was simpler final code.~~

### How to verify

To test the changes run updated tests or try to run analogous query to the one described in the issue in question.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

lbrdnk committed:
Update `LIKE` to handle fields and expressions

Previously only string could be used with like clause, ie. as second
argument of `:starts-with`, `:ends-with` or `:contains`. This commit modifies
translation of those functions, so fields or expressions can be
used in place of second argument.

lbrdnk committed:
Update tests

lbrdnk committed:
Make only expression or field args use `concat`

Using concat with values unnecessarily complects generated sql.

lbrdnk committed:
Extend mongo string matching capabilities

Make mongo filter compilation use `$regexMatch` for `:contains`,
`:starts-with` and `:ends-with` so expressions and fields
can be matched against.

lbrdnk committed:
Adjust mongo specific tests

lbrdnk committed:
Update druid

Druid's filters work with `:contains`, `:starts-with` and `:ends-with`,
but only if second argument is literal pattern. This commit updates
existing code, so expection is thrown if `pattern` has different form
than `[:value ...]`.

metamben commented on test/metabase/query_processor_test/filter_test.clj:
```diff
@@ -267,10 +267,28 @@
+                  [3 "The Apple Pan" 11 34.0406 -118.428 2]]
+                 (mt/formatted-rows :venues
+                                    (mt/run-mbql-query venues
+                                      {:filter   [:starts-with $name [:substring $name 1 3] {:case-sensitive false}]
```
Could we lowercase the expression to see case-insensitivity in action?

metamben committed:
Merge branch 'master' into 31284-contains-column-arg

metamben reviewed (commented):
Looks good, thanks! Could we keep compatibility with the previous versions and lowercase literals in Metabase? A few more tests would be nice to have for completeness.

lbrdnk committed:
Merge remote-tracking branch 'upstream/master' into 31284-contains-column-arg

lbrdnk committed:
Address review comments - sql

lbrdnk committed:
Address review comments - mongo

lbrdnk committed:
Use mbql `:concat` instead of `hx/call :concat`

Vertica, orcacle and sqlite handle concat differently than other
drivers. To overcome this limitation, `pre`, `post` and `arg` are
wrapped into mbql's concat expression which is then translated
with regards to the specifics of particular driver.

lbrdnk commented:
@metamben, I've modified the code according to your remarks. Changes have been successfully tested with h2, oracle, sqlite, vertica, postgres and sqlserver locally.

metamben committed:
Merge branch 'master' into 31284-contains-column-arg

lbrdnk committed:
Merge remote-tracking branch 'upstream/master' into 31284-contains-column-arg

lbrdnk commented:
@metamben, I've merged fresh master and resolved conflicts.

metamben committed:
Merge branch 'master' into 31284-contains-column-arg

lbrdnk committed:
Update according to the review feedback

lbrdnk commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -1062,33 +1062,34 @@
-  [value :- (s/constrained mbql.s/value #(string? (second %)) "string value"), f]
-  (update value 1 f))
+  ;; custom implementation? (Cam)
+  [:like (cond->> field (not case-sensitive) (hx/call :lower)) pattern])
```
Previously this was mentioned in only `generate-pattern`'s docstring. Now it can be found also in `like-clause`'s.

metamben committed:
Merge branch 'master' into 31284-contains-column-arg

guillaumemeyer commented:
Great improvement @lbrdnk !
Any idea in which version this update will land?

lbrdnk commented:
> Great improvement @lbrdnk ! Any idea in which version this update will land? Will test this asap
-- @guillaumemeyer

I don't know when this gets released, but maybe @darksciencebase will :)
